test/testers.jl

```diff
@@ -0,0 +1,816 @@
+- `rtol`: Relative tolerance for `isapprox`.
+- `atol`: Absolute tolerance for `isapprox`.
+- `testset_name`: Name to use for a testset in which all tests are evaluated.
+
+# Examples
+
+```julia
+x = randn()
+y = randn()  # will be Const
+for Tret in (Const, Duplicated, DuplicatedNoNeed), Tx in (Const, Duplicated)
+    test_forward(*, Tret, (x, Tx), y)
+end
+```
+"""
+function test_forward(
+    f,
+    ret_activity,
+    args...;
+    fdm=FiniteDifferences.central_fdm(5, 1),
+    fkwargs::NamedTuple=NamedTuple(),
+    rtol::Real=1e-9,
+    atol::Real=1e-9,
+    testset_name=nothing,
+)
+    call_with_copy(f, xs...) = deepcopy(f)(deepcopy(xs)...; deepcopy(fkwargs)...)
+    call_with_kwargs(f, xs...) = f(xs...; fkwargs...)
+    if testset_name === nothing
+        testset_name = "test_forward: $(f isa Const ? f.val : f) with return activity $ret_activity on $(args)"
+    end
+    @testset "$testset_name" begin
+        # format arguments for autodiff and FiniteDifferences
+        activities = map(auto_activity, (f, args...))
+        primals = map(x -> x.val, activities)
+        # call primal, avoid mutating original arguments
+        y = call_with_copy(primals...)
+        # call finitedifferences, avoid mutating original arguments
+        dy_fdm = _make_jvp_call(fdm, call_with_copy, ret_activity, y, activities)
+        # call autodiff, allow mutating original arguments
+        y_and_dy_ad = autodiff(Forward, call_with_kwargs, ret_activity, activities...)
+        if ret_activity <: Union{Duplicated,BatchDuplicated}
+            y_ad, dy_ad = y_and_dy_ad
+            # check primal agrees with primal function
+            test_approx(y_ad, y; atol, rtol)
+        elseif ret_activity <: Union{DuplicatedNoNeed,BatchDuplicatedNoNeed}
+            # check primal is not returned
+            @test length(y_and_dy_ad) == 1
+            dy_ad = y_and_dy_ad[1]
+        elseif ret_activity <: Const
+            # check Const activity returns an empty tuple
+            @test isempty(y_and_dy_ad)
+            dy_ad = ()
+        else
+            throw(ArgumentError("Unsupported return activity type: $ret_activity"))
+        end
+        if y isa Tuple
+            # check Enzyme and FiniteDifferences return the same number of derivatives
+            @test length(dy_ad) == length(dy_fdm)
+            # check all returned derivatives against FiniteDifferences
+            for (dy_ad_i, dy_fdm_i) in zip(dy_ad, dy_fdm)
+                test_approx(dy_ad_i, dy_fdm_i; atol, rtol)
+            end
+        else
+            test_approx(dy_ad, dy_fdm; atol, rtol)
+        end
+    end
+end
+
+"""
+    test_reverse(f, Activity, args...; kwargs...)
+
+Test `Enzyme.autodiff` of `f` in `Reverse`-mode against finite differences.
+
+`f` has the same constraints as `Enzyme.autodiff_thunk`. If it mutates one of its arguments,
+it must not also return that argument. If the return value is a struct, then all floating
```

Comment:
This limitation comes from us not being able to tell the difference between a shadow that is allocated in `augmented_primal` and the shadow of one of the inputs whose reference is returned by `augmented_primal`. Otherwise when constructing the `j′vp` call, we would take care to not add that input to the list of outputs. We could remove this limitation if the output is one of the inputs, but if instead the input is a field or element of the output, we could not detect that. So for clarity I've left this limitation as is.

Reply:
Sure!